crates/turbopack-ecmascript/src/tree_shake/util.rs

```diff
@@ -0,0 +1,219 @@
+        let old = self.is_read;
+        self.is_read = false;
+        p.visit_children_with(self);
+        self.is_read = old;
+    }
+
+    fn visit_pat_or_expr(&mut self, n: &PatOrExpr) {
+        if let PatOrExpr::Expr(e) = n {
+            let old = self.is_read;
+            self.is_read = false;
+            // Skip `visit_expr`
+            e.visit_children_with(self);
+            self.is_read = old;
+        } else {
+            n.visit_children_with(self);
+        }
+    }
+
+    fn visit_prop_name(&mut self, n: &PropName) {
+        if let PropName::Computed(..) = n {
+            n.visit_children_with(self);
+        }
+    }
+
+    noop_visit_type!();
+
+    visit_obj_and_computed!();
+}
+
+#[derive(Default)]
+pub(crate) struct CapturedIdCollector {
+    reads: Vec<Id>,
+    writes: Vec<Id>,
+    is_nested: bool,
+    is_read: bool,
+}
+
+impl Visit for CapturedIdCollector {
+    fn visit_block_stmt_or_expr(&mut self, n: &BlockStmtOrExpr) {
+        let old = self.is_nested;
+        self.is_nested = true;
+        n.visit_children_with(self);
+        self.is_nested = old;
+    }
+
+    fn visit_constructor(&mut self, n: &Constructor) {
+        let old = self.is_nested;
+        self.is_nested = true;
+        n.visit_children_with(self);
+        self.is_nested = old;
+    }
+
+    fn visit_expr(&mut self, e: &Expr) {
+        let old = self.is_read;
+        self.is_read = true;
+        e.visit_children_with(self);
+        self.is_read = old;
+    }
+
+    fn visit_function(&mut self, n: &Function) {
+        let old = self.is_nested;
+        self.is_nested = true;
+        n.visit_children_with(self);
+        self.is_nested = old;
+    }
+
+    fn visit_ident(&mut self, n: &Ident) {
+        if self.is_nested {
+            if self.is_read {
+                if self.reads.iter().any(|id| id == &n.to_id()) {
+                    return;
+                }
+                self.reads.push(n.to_id());
+            } else {
+                if self.writes.iter().any(|id| id == &n.to_id()) {
+                    return;
+                }
+                self.writes.push(n.to_id());
+            }
+        }
+    }
+
+    fn visit_pat(&mut self, p: &Pat) {
+        let old = self.is_read;
+        self.is_read = false;
+        p.visit_children_with(self);
+        self.is_read = old;
+    }
+
+    fn visit_pat_or_expr(&mut self, n: &PatOrExpr) {
+        if let PatOrExpr::Expr(e) = n {
+            let old = self.is_read;
+            self.is_read = false;
+            // Skip `visit_expr`
+            e.visit_children_with(self);
+            self.is_read = old;
+        } else {
+            n.visit_children_with(self);
+        }
+    }
+
+    fn visit_prop_name(&mut self, n: &PropName) {
+        if let PropName::Computed(..) = n {
+            n.visit_children_with(self);
+        }
+    }
+
+    noop_visit_type!();
+
+    visit_obj_and_computed!();
+}
+
+/// Returns `(read, write)`
```

Comment:
This is used in 3 functions. Maybe we can just create a struct for that `struct ReadWriteIds { reads: Vec<Id>, writes: Vec<Id> }`